### ISSUE TITLE
Grammar reference

### ISSUE THREAD

**Issue by teucer** [bambinos/formulae#66]
I could not find any reference on the grammar that is supported. How does it compare to [formulaic](https://matthewwardrop.github.io/formulaic/basic/grammar/)?

**Comment by tomicapretto** [comment]
You're right there's no formal grammar reference. The closest we have is the [Getting started section in our docs](https://bambinos.github.io/formulae/notebooks/getting_started.html).

Nevertheless, formulae and formulaic should agree in the design matrices that are constructed from the same formula. The internals are different, but the results should match. 

One of the most relevant differences is that formulae implements the `|` operator to construct design matrices for random effects, while formulae does not (yet).

**Comment by teucer** [comment]
Thank you.

How does it compare to R? We are in the process of migrating an R package to python, it would be useful to know them.

**Comment by tomicapretto** [comment]
Could you tell me which features you're interested in? That could help to point you to relevant differences.

Formulae is very similar to model formulas in R. It still lacks the double pipe `||` operator, but maybe it is not relevant for you.

**Comment by teucer** [comment]
I was wondering about diff or lag operators for example. It is not clear to me how vector operators are handled.

**Comment by tomicapretto** [comment]
I think this is going to be clearer if you share a concrete R example with both the input and the output. In particular, how do you use `diff()` or `lag()` in a R formula?

**Comment by teucer** [comment]
Some thing like below:

`log(y) ~ lag(y) + diff(log(x))`

As you can see diff and lag apply to the whole 'vector', they are not scalar functions.

**Comment by tomicapretto** [comment]
@teucer could you share a reproducible example? That's what I mean with both input and output. I understand the formula, but I don't know what is the expected behavior from that formula. 

What I would like to have is the following

* Sample input data
* Sample formula
* How the formula is used (e.g. is it used within `lm()` or within which modeling function?)
* Sample output


As far as I know, the result of `lag(x)` is of length `n`, but the result of `diff(x)` is of length `n - 1`.

```r
> x <- 1:5
> length(lag(x))
# [1] 5
> length(diff(x))
# [1] 4
```

**Comment by teucer** [comment]
Sorry for the lengthy discussion, I am not the author of the package that we are trying to convert. It seems that people have implemented their own `lag` and `diff` functions. 

Now,  it would be useful if we could pass our own additional transformations. I think you are doing that with Environment. What about passing a dictionary of transformations? It might be the case that we are defining them in another file. 

```python
# this could be defined in an another file and the dictionary imported
def myfun(x):
    return x + 1

trans_dict = {"myfun":  myfun}
 
dm = design_matrices("y1 ~ myfun(x)", data, transformations=trans_dict)
```

**Comment by tomicapretto** [comment]
Thanks for the comments clarifying my question! 

1. Formulae should work fine with array valued functions. 
2. The problem with the first `lag()` function is that it returns `np.nan`. Currently, Formulae either raises an error or drops rows containing missing values. We should add a case where we just want to keep missing values. See 

https://github.com/bambinos/formulae/blob/38d7f42f7d642d7ab86e87b5d6d138efb6394785/formulae/matrices.py#L524-L533

3. At the moment, it is not possible to pass a dictionary of transformations to `design_matrices()`. Maybe you can modify the following dictionary 

https://github.com/bambinos/formulae/blob/38d7f42f7d642d7ab86e87b5d6d138efb6394785/formulae/transforms.py#L401-L412

which is what Formulae consumes when looking for internal transformations. However I don't think this is a clean solution. 

4. How many transformations do you have? If they are few, you could just do `from .utils import lag, diff,  etc` and that would solve the problem.

**Comment by tomicapretto** [comment]
It would be good if you could try to work on a PR for 2 (allowing to keep missing values) and 3, allowing to pass a dictionary with transformations. 

For 3, have a look at 

https://github.com/bambinos/formulae/blob/38d7f42f7d642d7ab86e87b5d6d138efb6394785/formulae/matrices.py#L513

and https://github.com/bambinos/formulae/blob/master/formulae/environment.py

and feel free to ask questions. 

As a pointer, I would create a new Environment holding the transforms you want and./or use `.with_outer_namespace()`.

**Comment by tomicapretto** [comment]
@teucer thanks for the contributions y #69 and #70. Is the development version working for your needs now?

**Comment by tomicapretto** [comment]
The development version as a lot of breaking changes. I want to double check a couple of things before doing a release.
Can you install from github in the meantime?

**Comment by teucer** [comment]
Thank you for the support. Looking forward to the new release. 

PS: I would be happy to further contribute if you have further tasks. If you add tasks and tag them (e.g. need contributor), it would be easier for me to pickup and send PRs.

**Comment by tomicapretto** [comment]
@teucer there's a new release now :)
